viadot/sources/sap_bw.py

```diff
@@ -0,0 +1,120 @@
+            mdx_query (str, optional): The MDX query to be passed to connection. Defaults to None.
+
+        Returns:
+            all_available_columns: List of all available columns in the source table.
+        """
+        conn = self.get_connection()
+        query = textwrap.wrap(
+            mdx_query, width=75
+        )  # width = 75, to properly split mdx query into substrings passed to SAP object creation function
+
+        properties = conn.call("RSR_MDX_CREATE_STORED_OBJECT", COMMAND_TEXT=query)
+        datasetid = properties["DATASETID"]
+
+        if properties["RETURN"]["MESSAGE"] == "":
+            get_axis_info = conn.call(
+                "RSR_MDX_GET_AXIS_INFO", DATASETID=datasetid
+            )  # listing all of available columns and metrics
+            cols = get_axis_info["AXIS_DIMENSIONS"]
+
+            all_available_columns = [x["DIM_UNAM"] for x in cols]
+        else:
+            all_available_columns = []
+            self.logger.error(properties["RETURN"]["MESSAGE"])
+
+        return all_available_columns
+
+    def get_output_data(self, mdx_query: str = None) -> dict:
+        """
+        Function to generate the SAP output dataset from MDX query.
+
+        Args:
+            mdx_query (str, optional): The MDX query to be passed to connection. Defaults to None.
+
+        Returns:
+            query_output: SAP output dictionary in JSON format that contains data rows and column headers.
+
+                Example output:
+
+                    {
+                    "RETURN": {
+                        "TYPE": "",
+                        "ID": "",
+                        "NUMBER": "000",
+                        "MESSAGE": "",...
+                    },...
+                    "DATA": [
+                        {
+                        "COLUMN": 0,
+                        "ROW": 0,
+                        "DATA": "VELUX Deutschland GmbH",
+                        "VALUE_DATA_TYPE": "CHAR",
+                        "CELL_STATUS": ""
+                        },...
+                    ],
+                    "HEADER": [
+                        {
+                        "COLUMN": 0,
+                        "ROW": 0,
+                        "DATA": "[0COMP_CODE].[LEVEL01].[DESCRIPTION]",
+                        "VALUE_DATA_TYPE": "CHAR",
+                        "CELL_STATUS": ""
+                        },...
+                    ]
+                    }
+
+        """
+        conn = self.get_connection()
+        query = textwrap.wrap(
+            mdx_query, 75
+        )  # width = 75, to properly split mdx query into substrings passed to SAP object creation function
+        properties = conn.call("RSR_MDX_CREATE_OBJECT", COMMAND_TEXT=query)
+
+        datasetid = properties["DATASETID"]
+        query_output = conn.call("RSR_MDX_GET_FLAT_DATA", DATASETID=datasetid)
+
+        # if properties["RETURN"]["MESSAGE"] != "":
+        #     self.logger.error(properties["RETURN"]["MESSAGE"])
```

Comment:
Remove commented lines if they're not needed.

Reply:
cleaned ;)